### PR TITLE
HHH-18839 Validate property type of `@IdClass` at bind time

### PR DESCRIPTION
**Pull request by quaff**
----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

https://hibernate.atlassian.net/browse/HHH-18839

### CODE REVIEW

**Comment by quaff** [comment]
Ready for review, need your approval to run CI @gavinking

**Review by gavinking — CHANGES_REQUESTED** [review]
Thanks @quaff, please see my comments.

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java [hunk 1]
```diff
@@ -852,11 +853,38 @@ private static void processIdClassElements(
...
+		if ( value.indexOf( '.' ) < 0 ) {
+			String wrapper = switch ( value ) {
+				case "int" -> "Integer";
+				case "char" -> "Character";
+				default -> Character.toUpperCase( value.charAt( 0 ) ) + value.substring( 1 );
+			};
+			return "java.lang." + wrapper;
+		}
+		return value;
+	}
```
[the remark this answers is not in the document]
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java]: done.

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java [hunk 2]
```diff
@@ -852,11 +853,38 @@ private static void processIdClassElements(
...
 			classElements.set( i, entityPropertyData );  //this works since they are in the same order
 		}
 	}
 
+	private static boolean hasCompatibleType( String typeNameInIdClass, String typeNameInEntityClass ) {
+		if ( typeNameInIdClass.equals( typeNameInEntityClass ) ) {
+			return true;
+		}
+		return toWrapper( typeNameInIdClass ).equals( typeNameInEntityClass ) || typeNameInIdClass.equals(
+				toWrapper( typeNameInEntityClass ) );
```
[the remark this answers is not in the document]
Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java]: done.

File: hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java [hunk 3]
```diff
@@ -852,11 +853,38 @@ private static void processIdClassElements(
...
+				}
 			}
 			classElements.set( i, entityPropertyData );  //this works since they are in the same order
 		}
 	}
 
+	private static boolean hasCompatibleType(String typeNameInIdClass, String typeNameInEntityClass) {
+		return typeNameInIdClass.equals( typeNameInEntityClass )
+				|| canonicalize( typeNameInIdClass ).equals( typeNameInEntityClass )
+				|| typeNameInIdClass.equals( canonicalize( typeNameInEntityClass ) );
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java:871]: I still think this code is equivalent to `canonicalize( typeNameInIdClass ).equals( canonicalize( typeNameInEntityClass ) )`. Is that not the case?

Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java:871]: Yes, but that is less efficient, `typeNameInIdClass.equals( typeNameInEntityClass )` will be hit in most cases.

Comment (quaff) [reply hibernate-core/src/main/java/org/hibernate/boot/model/internal/EmbeddableBinder.java:871]: > Yes, but that is less efficient, `typeNameInIdClass.equals( typeNameInEntityClass )` will be hit in most cases.

@gavinking WDYT?

**Review by gavinking — APPROVED** [review]
OK, this looks OK now, thanks!

**Pull request merged**